services/CurrentTime/include/ble-service-current-time/CurrentTimeService.h

```diff
@@ -0,0 +1,178 @@
+     */
+    CurrentTimeService(BLE &ble, events::EventQueue &event_queue) :
+            _ble(ble),
+            _event_queue(event_queue),
+            _current_time_char(GattCharacteristic::UUID_CURRENT_TIME_CHAR, &_current_time) {
+        memset(&_current_time, 0, sizeof(_current_time));
+    }
+
+    ble_error_t init() {
+        GattCharacteristic *charTable[] = {&_current_time_char};
+        GattService currentTimeService(GattService::UUID_CURRENT_TIME_SERVICE, charTable, 1);
+
+        _current_time_char.setReadAuthorizationCallback(this, &CurrentTimeService::onDataRead);
+        _current_time_char.setReadAuthorizationCallback(this, &CurrentTimeService::onDataRead);
+
+        ble_error_t bleError = _ble.gattServer().addService(currentTimeService);
+
+        return bleError;
+    }
+
+    time_t get_current_time() {
+        time_t t = time(NULL);
+
+        return _current_time_offset + t;
+    }
+
+    void set_time(time_t new_time) {
+        time_t t = time(NULL);
+
+        _current_time_offset = t - new_time;
+    }
+
+private:
+    void set_current_time() {
+        timeval tv{};
+        gettimeofday(&tv, NULL);
+        // printf("%u\r\n", (unsigned int) tv.tv_sec);
+
+        struct tm *current_time;
+        time_t local_time = tv.tv_sec + _current_time_offset;
+        current_time = localtime(&local_time);
+
+        _current_time.year    = current_time->tm_year + 1900;
+        _current_time.month   = current_time->tm_mon  + 1;
+        _current_time.day     = current_time->tm_mday;
+        _current_time.hours   = current_time->tm_hour;
+        _current_time.minutes = current_time->tm_min;
+        _current_time.seconds = current_time->tm_sec;
+        _current_time.weekday = (current_time->tm_wday == 0) ? 7 : current_time->tm_wday;
+        // 1/256th of seconds not supported
+        _current_time.fractions256 = 0;
+    }
+
+    void onDataRead(GattReadAuthCallbackParams *read_request) {
+        set_current_time();
+
+        _ble.gattServer().write(_current_time_char.getValueHandle(),
+                                reinterpret_cast<const uint8_t *>(&_current_time),
+                                sizeof(_current_time));
+
+        read_request->authorizationReply = AUTH_CALLBACK_REPLY_SUCCESS;
+    }
+
+    void onDataWritten(GattWriteAuthCallbackParams *write_request) {
+        const uint8_t *ct = (write_request->data);
+
+        struct tm *current_time;
+
+        current_time->tm_year = (*ct | (*(ct + 1) << 8)) - 1900;
+        ct += 2;
+        current_time->tm_mon  =  *ct++ - 1;
+        current_time->tm_mday =  *ct++;
+        current_time->tm_hour =  *ct++;
+        current_time->tm_min  =  *ct++;
+        current_time->tm_sec  =  *ct++;
+        current_time->tm_wday =  (*ct == 7 ? 0 : *ct);
+
+        time_t remote_time = mktime(current_time);
+
+        set_time(remote_time);
+    }
+
+    BLE &_ble;
+    events::EventQueue &_event_queue;
+
+    CurrentTime _current_time;
```

Comment:
Two options to make sure the structure as a proper layout: 
- Use `MBED_PACKED`
- Wrap this into an union with an array of byte of the size wanted then static assert the size of the union.